### PR TITLE
New features: Keyboard controls, callbacks and destroy with initial markup

### PR DESCRIPTION
**Pull request by znk**
Hello there, 
Thanks for your amazing and lightweight carousel. Because. I needed some features over last weeks without time diving into, I decided to finally add them! I also updated the corresponding documentation.

Please, take time to review and criticize PR. I will be quick to make changes to see them part of Siema!

### CODE REVIEW

**Comment by znk** [comment]
Callbacks were named `afterPrev` instead of `onPrev` for clarity. And also, I was thinking to add `beforePrev` in advance. It can be useful to trigger an action before displaying the next slide, such as lazy loading content.
```js
  // Go to previous slide
  prev(howManySlides = 1) {
    if (this.currentSlide === 0 && this.config.loop) {
      this.currentSlide = this.innerElements.length - this.perPage;
    }
    else {
      this.currentSlide = Math.max(this.currentSlide - howManySlides, 0);
    }
    this.config.beforePrev.call(this);
    this.slideToCurrent();
    this.config.onChange.call(this);
    this.config.afterPrev.call(this);
  }
```

**Comment by znk** [comment]
#26 destroy with initial markup, need testing for edge cases.

**Comment by pawelgrzybek** [comment]
Good morning!

O wow! Whoever you are, wherever you live I will meet you one day and have a beer with you my friend! Amazing contribution, amazing! Top quality code and really great solutions. You can be assured that your PR will be merged with Siema (if not entirely, at least big part of it). I will test everything and leave some helpful comments if needed. I will do it this weekend...

Can't thank you enough for this amazing work that you've done!
Have a great day ⭐️

**Comment by xzentorzx** [comment]
Hey there, what's the state on this? 
I could really use that reversion of the markup ;)

Thanks for a really neat library (and addition 😊 )

**Comment by pawelgrzybek** [comment]
I did a quick review of it. `destoy()` method + corresponding callback will become part of Siema. It requires a little bit more testing but I will do it very shortly guys. Thanks a ton for involvement you all.

**Comment by znk** [comment]
The best workaround is still to store the initial markup in a DocumentFragment object during `init()`, and then revert it back on `destroy()`.

**Comment by znk** [comment]
Discarding extra (async) items is troublesome and counter intuitive, to my opinion. Siema' `destroy()` should only remove itself without regards to what it holds. Number two :+1:

**Comment by xzentorzx** [comment]
That's also what I would suggest.

**Comment by znk** [comment]
Let's merge. `destroy()` may need to be ... polished up :sunglasses: , but lets wait for actual feedback on some eventual issue.

**Comment by pawelgrzybek** [comment]
I will do it this week @znk. Really need a little bit of testing (at list quickly test some browser compatibility). Didn't have a chance to to do it last weekend — sorry about it. I'm considering optional argument for this method to decide about reverting markup to initial state or keeping it and just remove listeners. I see a use cases for both scenarios. 

Like so...

```js
destro(restoreToInitialMarkup = false) {
  // remove all listeners

  if (restoreToInitialMarkup) {
    // revert to initial markup
  }
}
```

Any thoughts @zkn ?

File: README.md [hunk 1]
```diff
@@ -53,6 +53,9 @@ new Siema({
   loop: false,
   onInit: function(){},
   onChange: function(){},
+  afterPrev: function() {},
+  afterNext: function() {},
+  onDestroy: function() {}
```
Comment (pawelgrzybek) [inline README.md]: We have to ditch `afterPrev()` and `afterNext()` from this PR. In current implementation it invokes actually straight before `next()` and `prev()`, not after it. Doing before and after is a bit more complicated as it needs to listen for `transitionstart` and `transitionend` events on instance's `sliderFrame`. Let's keep it for future versions of Siema.

The `onDestory()` callback — perfecto :)

Don't worry about changing that, I will sort it out dude.

File: README.md [hunk 2]
```diff
@@ -66,6 +69,9 @@ new Siema({
 - `loop` - (boolean) loop the slides around
 - `onInit` - (function) runs immediately after first initialization
 - `onChange` - (function) runs after slide change
+- `afterPrev` - (function) runs after going to previous slide
+- `afterNext` - (function) runs after going to next slide
+- `onDestroy` - (function) runs after removing all active listeners
```
Comment (pawelgrzybek) [inline README.md]: We have to ditch afterPrev() and afterNext() from this PR. In current implementation it invokes actually straight before next() and prev(), not after it. Doing before and after is a bit more complicated as it needs to listen for transitionstart and transitionend events on instance's sliderFrame. Let's keep it for future versions of Siema.

The onDestory() callback — perfecto :)

File: src/siema.js [hunk 3]
```diff
@@ -56,8 +56,11 @@ export default class Siema {
       draggable: true,
       threshold: 20,
       loop: false,
-      onInit: function() {},
-      onChange: function() {},
+      onInit: () => {},
+      onChange: () => {},
+      afterPrev: () => {},
+      afterNext: () => {},
+      onDestroy: () => {}
```
Comment (pawelgrzybek) [inline src/siema.js]: We have to ditch afterPrev() and afterNext() from this PR. In current implementation it invokes actually straight before next() and prev(), not after it. Doing before and after is a bit more complicated as it needs to listen for transitionstart and transitionend events on instance's sliderFrame. Let's keep it for future versions of Siema.

The onDestory() callback — perfecto :)

File: src/siema.js [hunk 4]
```diff
@@ -146,6 +149,7 @@ export default class Siema {
     }
     this.slideToCurrent();
     this.config.onChange.call(this);
+    this.config.afterPrev.call(this);
```
Comment (pawelgrzybek) [inline src/siema.js]: needs to go away :(

File: src/siema.js [hunk 5]
```diff
@@ -159,6 +163,7 @@ export default class Siema {
     }
     this.slideToCurrent();
     this.config.onChange.call(this);
+    this.config.afterNext.call(this);
```
Comment (pawelgrzybek) [inline src/siema.js]: needs to go away :(

File: src/siema.js [hunk 6]
```diff
@@ -300,5 +305,6 @@ export default class Siema {
     this.selector.removeEventListener('mouseup', this.mouseupHandler);
     this.selector.removeEventListener('mouseleave', this.mouseleaveHandler);
     this.selector.removeEventListener('mousemove', this.mousemoveHandler);
+    this.config.onDestroy.call(this);
```
Comment (pawelgrzybek) [inline src/siema.js]: perfecto :)

**Comment by znk** [comment]
@pawelgrzybek thanks for reviewing, I second your idea on `destroy(restoreToInitialMarkup = false)`. What alternative to these callbacks? ATM I handle it by keeping on eye on `onChange()` and `currentSlide` to get the direction and lazyload in advance, but that's a bit rudimentary...

File: src/siema.js [hunk 7]
```diff
@@ -16,7 +16,7 @@ export default class Siema {
     this.init();
 
     // Bind all event handlers for referencability
-    ['resizeHandler', 'touchstartHandler', 'touchendHandler', 'touchmoveHandler', 'mousedownHandler', 'mouseupHandler', 'mouseleaveHandler', 'mousemoveHandler'].forEach(method => {
+    ['resizeHandler', 'touchstartHandler', 'touchendHandler', 'touchmoveHandler', 'mousedownHandler', 'mouseupHandler', 'mouseleaveHandler', 'mousemoveHandler', 'keydownHandler'].forEach(method => {
```
Comment (pawelgrzybek) [inline src/siema.js]: Sorry @znk but hijacking `keydown` event on document body may be dangerous. It can harm a lot core functionality on the project and it's tricky from accessibility point of view as well. I prefer users who are conscious what they do, to do it by themselves (if they really have to). Personally I consider it as an anti-pattern on majority of cases. The only reasonable use case is having a full page carousel and nothing else on the page (like a presentation mode).

File: src/siema.js [hunk 8]
```diff
@@ -42,6 +42,9 @@ export default class Siema {
       this.selector.addEventListener('mouseup', this.mouseupHandler);
       this.selector.addEventListener('mouseleave', this.mouseleaveHandler);
       this.selector.addEventListener('mousemove', this.mousemoveHandler);
+
+      // Keyboard events
+      document.addEventListener('keydown', this.keydownHandler);
```
Comment (pawelgrzybek) [inline src/siema.js]: Sorry @znk but hijacking `keydown` event on document body may be dangerous. It can harm a lot core functionality on the project and it's tricky from accessibility point of view as well. I prefer users who are conscious what they do, to do it by themselves (if they really have to). Personally I consider it as an anti-pattern on majority of cases. The only reasonable use case is having a full page carousel and nothing else on the page (like a presentation mode).

File: src/siema.js [hunk 9]
```diff
@@ -294,6 +297,17 @@ export default class Siema {
...
+    switch (e.key) {
+      case "ArrowLeft":
+        this.prev();
+        break;
+
+      case "ArrowRight":
+        this.next();
+        break;
+    }
+  }
```
Comment (pawelgrzybek) [inline src/siema.js]: Sorry @znk but hijacking `keydown` event on document body may be dangerous. It can harm a lot core functionality on the project and it's tricky from accessibility point of view as well. I prefer users who are conscious what they do, to do it by themselves (if they really have to). Personally I consider it as an anti-pattern on majority of cases. The only reasonable use case is having a full page carousel and nothing else on the page (like a presentation mode).

File: src/siema.js [hunk 10]
```diff
@@ -305,6 +319,7 @@ export default class Siema {
     this.selector.removeEventListener('mouseup', this.mouseupHandler);
     this.selector.removeEventListener('mouseleave', this.mouseleaveHandler);
     this.selector.removeEventListener('mousemove', this.mousemoveHandler);
+    document.removeEventListener('keydown', this.keydownHandler);
```
Comment (pawelgrzybek) [inline src/siema.js]: Sorry @znk but hijacking `keydown` event on document body may be dangerous. It can harm a lot core functionality on the project and it's tricky from accessibility point of view as well. I prefer users who are conscious what they do, to do it by themselves (if they really have to). Personally I consider it as an anti-pattern on majority of cases. The only reasonable use case is having a full page carousel and nothing else on the page (like a presentation mode).

**Comment by pawelgrzybek** [comment]
@znk 

> @pawelgrzybek thanks for reviewing, I second your idea on destroy(restoreToInitialMarkup = false).

Do you like this idea? Personally I can se a use case for both scenarios...

> What alternative to these callbacks (prev & next) ?

Maybe simply `onPrev()` and `onNext()` to keep it simple and have an access to direction of sliding? I will care about the before and after event later on — as I said this one requires a bit more work and much more browser testing...

File: src/siema.js (hunk 10 above)
Comment (znk) [inline src/siema.js]: True, thanks for feedback. I do agree and will continue to manage it outside.

Comment (pawelgrzybek) [inline src/siema.js]: It makes more sense @znk. It should be controlled from app level. At least optinal, but enabling thing like this out of the box may harm a lot.

**Comment by znk** [comment]
> Really need a little bit of testing (at list quickly test some browser compatibility). 
As for testing, DocumentFragment and `cloneNode()` are part of DOM1 and DOM2, roughly both around 2000. Need some help to these E2E tests?

> Do you like this idea? Personally I can se a use case for both scenarios...
Yes, that is a good solution.

> Maybe simply onPrev() and onNext() to keep it simple and have an access to direction of sliding?
Fine, I agree too. It does make it easier to handle lazyloaded content :+1: 

In short:
- keyboard controls are ditched as they should be managed out of Siema for stated reasons
- simple and efficient callbacks `onPrev()` and `onNext()` will be shipped :relieved:
- initial markup on `destroy()` will be made optional (:punch:)

@pawelgrzybek What about the callback `onDestroy()`?

**Comment by pawelgrzybek** [comment]
> What about the callback onDestroy()?

We will see this one on 1.4.0 as well :)

I just created a new branch 1.4.0. The part of your PR that is going to be merged with next version of Siema is already there. As said before — need a bit of testing and few other cool features to this release. Should land shortly.

If you can help with e2e later on (lets say 1.5.0) I will owe you a dinner and tons of beer not a single pint as promised before :) 🍻

**Comment by pawelgrzybek** [comment]
Nearly there. Just tested everything on IE10. Your PR will be added with just a tiny more complexity to with optional argument to restore Initial markup.

http://codepen.io/pawelgrzybek/pen/KWYLKx

**Comment by znk** [comment]
@pawelgrzybek i did saw the pen but did not have time to comment. it seemed better like you did, to use `innerHTML`. i see your activity on 1.4.0... oh boy,  you are doing many wonders :)

**Comment by pawelgrzybek** [comment]
Thanks for a kind words. I will release 1.4.0 very soon. The `destroy()` method is based on your commit, the only difference is using `innerHTML` instead of cloning node. Rest of it is yours man.
